File: phoneBooks/src/phoneBooks/PhoneBook.java
package phoneBooks;

import java.util.ArrayList;
import java.util.List;

public class PhoneBook {
    private List<Contact> contacts = new ArrayList<>();

    public List<Contact> getContacts() {
        return contacts;
    }

    public void addContact(Contact contact) {
        if (contacts.contains(contact)) {
            updateContact(contact);
        } else {
            contacts.add(contact);
        }
    }

    public void updateContact(Contact contact) {
        int i = contacts.indexOf(contact);
        if (i != -1) {
            contacts.set(i, contact);
        } else {
            addContact(contact);
        }
    }

    public boolean removeContact(Contact contact) {
        if (contacts.contains(contact)) {
            contacts.remove(contact);
            return true;
        }
        return false;
    }

    public Contact searchByTelephone(String phone) {
        for (Contact c : contacts) {
            if(c.getTelephone().equals(phone)) {
                return c;
            }
        }
        return null;
    }

    public Contact searchByName(String name) {
        for (Contact c : contacts) {
            if (c.getFullName().equals(name)) {
                return c;
            }
        }
        return null;
    }

    @Override
    public String toString(){
        StringBuilder sb = new StringBuilder();
        for (Contact c : contacts) {
            sb.append(c.toString()).append("\n");
        }
        return sb.toString();
    }

    @Override
    public boolean equals(Object o) {
        PhoneBook b = (PhoneBook) o;
        List<Contact> bList = b.getContacts();
        if (contacts.size() != bList.size()) {
            return false;
        }
        for (Contact c : bList) {
            if (!contacts.contains(c)) {
                return false;
            }
        }
        return true;
    }
}

File: phoneBook/phoneBook/PhoneBook.cs
using phoneBook.interfaceBoks;

namespace phoneBook.phoneBook
{
    public class PhoneBook : IPhoneBook
    {
        private List<Contact> contacts = new List<Contact>();
        public List<Contact> Contacts
        {
            get => contacts;
        }

        public virtual bool Add(Contact contact)
        {
            if (!contacts.Contains(contact))
            {
                contacts.Add(contact);
                return true;
            }
            return false;
        }

        public bool Update(Contact contact)
        {
            Contact? c = SearchByName(contact.FullName);
            if (c != null
                && !c.Equals(contact)
                && IsLooksLike(c, contact))
            {
                if (Remove(c))
                {
                    return Add(contact);
                }
            }
            return false;
        }

        private bool IsLooksLike(Contact c1, Contact c2)
        {
            foreach (string s in c1.Phone.Telephones)
            {
                if (c2.Phone.IsPhone(s))
                {
                    return true;
                }
            }
            return false;
        }

        public virtual bool Remove(Contact contact)
        {
            return contacts.Remove(contact);
        }

        public Contact? SearchByTelephone(string t)
        {
            foreach (Contact item in contacts)
            {
                if (item.Phone.IsPhone(t))
                {
                    return item;
                }
            }
            return null;
        }

        public Contact? SearchByName(string name)
        {
            foreach (Contact item in contacts)
            {
                if (item.FullName.Equals(name))
                {
                    return item;
                }
            }
            return null;
        }

        public override bool Equals(object? obj)
        {
            if (obj != null && this.GetType()
                .Equals(obj.GetType()))
            {
                PhoneBook pB = (PhoneBook)obj;
                foreach (Contact c in pB.Contacts)
                {
                    if (!contacts.Contains(c))
                    {
                        Console.WriteLine("-----3");
                        return false;
                    }
                }
            }
            return true;
        }

        public override int GetHashCode()
        {
            int res = 0;
            foreach (Contact item in contacts)
            {
                res += item.GetHashCode();
            }
            return res;
        }

        public override string ToString()
        {
            string str = string.Empty;
            foreach (Contact item in contacts)
            {
                str += item + "\n";
            }
            return str;
        }

        public string ToFile()
        {
            string str = string.Empty;
            foreach (Contact c in contacts)
            {
                str += c.FullName + ":" + c.Phone.PhonesToFile() + "\n";
            }
            return str;
        }

    }
}